core/src/main/scala/kafka/coordinator/transaction/ProducerIdManager.scala

```diff
@@ -75,87 +61,179 @@ trait ProducerIdGenerator {
       zkWriteComplete = succeeded
 
-      if (zkWriteComplete)
-        info(s"Acquired new producerId block $currentProducerIdBlock by writing to Zk with path version $version")
+      if (zkWriteComplete) {
+        logger.info(s"Acquired new producerId block $newProducerIdBlock by writing to Zk with path version $version")
+        return newProducerIdBlock
+      }
     }
+    throw new IllegalStateException()
   }
+}
 
-  private def checkProducerIdBlockZkData(zkClient: KafkaZkClient, path: String, expectedData: Array[Byte]): (Boolean, Int) = {
-    try {
-      val expectedPidBlock = ProducerIdManager.parseProducerIdBlockData(expectedData)
-      zkClient.getDataAndVersion(ProducerIdBlockZNode.path) match {
-        case (Some(data), zkVersion) =>
-          val currProducerIdBLock = ProducerIdManager.parseProducerIdBlockData(data)
-          (currProducerIdBLock == expectedPidBlock, zkVersion)
-        case (None, _) => (false, -1)
-      }
-    } catch {
-      case e: Exception =>
-        warn(s"Error while checking for producerId block Zk data on path $path: expected data " +
-          s"${new String(expectedData, StandardCharsets.UTF_8)}", e)
-        (false, -1)
-    }
+class ZkProducerIdManager(brokerId: Int,
+                          zkClient: KafkaZkClient) extends ProducerIdGenerator with Logging {
+
+  private var currentProducerIdBlock: ProducerIdsBlock = ProducerIdsBlock.EMPTY
+  private var nextProducerId: Long = -1L
+
+  // grab the first block of producerIds
+  this synchronized {
+    getNewProducerIdBlock()
+    nextProducerId = currentProducerIdBlock.producerIdStart
+  }
+
+  private def getNewProducerIdBlock(): Unit = {
+    currentProducerIdBlock = ZkProducerIdManager.getNewProducerIdBlock(brokerId, zkClient, this)
   }
 
   def generateProducerId(): Long = {
     this synchronized {
       // grab a new block of producerIds if this block has been exhausted
-      if (nextProducerId > currentProducerIdBlock.blockEndId) {
+      if (nextProducerId > currentProducerIdBlock.producerIdEnd) {
         getNewProducerIdBlock()
-        nextProducerId = currentProducerIdBlock.blockStartId + 1
+        nextProducerId = currentProducerIdBlock.producerIdStart + 1
       } else {
         nextProducerId += 1
       }
-
       nextProducerId - 1
     }
   }
+}
+
+class ProducerIdManager(brokerId: Int,
+                        brokerEpochSupplier: () => Long,
+                        controllerChannel: BrokerToControllerChannelManager,
+                        maxWaitMs: Int) extends ProducerIdGenerator with Logging {
+
+  this.logIdent = "[ProducerId Manager " + brokerId + "]: "
+
+  private val nextProducerIdBlock = new ArrayBlockingQueue[Try[ProducerIdsBlock]](1)
+  private val requestInFlight = new AtomicBoolean(false)
+  private var currentProducerIdBlock: ProducerIdsBlock = ProducerIdsBlock.EMPTY
+  private var nextProducerId: Long = -1L
+
+  override def generateProducerId(): Long = {
```

Comment:
It would be better to use a condition variable rather than a blocking queue here. Consider the situation where the latest RPC call got an error. In that case, you want to deliver the error to all waiters, not just to one of them. You could do that with `notifyAll()`. You can't really do it with a blocking queue because there's 1 consumer and 1 producer.

Removing the blocking queue also avoids using extra locks and so on. We only really need one lock and one condition variable here, I think. (wait / notify should be fine here, no need for ReentrantLock since we don't need multiple cvars).

Reply:
To enlarge on the previous comment a bit: I guess the reason why I consider a condition variable to be simpler is that it involves fewer locks. A blocking queue has its own lock separate from the the lock in the manager, which can be awkward.

What kind of behavior do you want on error? Do you want to deliver a given error to all waiters, or just to a single waiter? It seems like you might need to deliver it to all of them, to prevent huge pile-ups in cases where there are such errors.

Reply:
Let's do this in a follow-on